src/transport_ofi.c

```diff
@@ -1107,37 +1115,103 @@ int query_for_fabric(struct fabric_info *info)
     return ret;
 }
 
+static int shmem_transport_ofi_ctx_init(shmem_transport_ctx_t *ctx, int id)
+{
+    int ret = 0;
+
+    struct fi_cntr_attr cntr_attr = {0};
+    cntr_attr.events   = FI_CNTR_EVENTS_COMP;
+    /* Set FI_WAIT based on the put and get polling limits defined above */
+    if (shmem_transport_ofi_put_poll_limit < 0) {
+        cntr_attr.wait_obj = FI_WAIT_NONE;
+    } else {
+        cntr_attr.wait_obj = FI_WAIT_UNSPEC;
+    }
+
+    struct fabric_info* info = &shmem_transport_ofi_info;
+    info->p_info->ep_attr->tx_ctx_cnt = FI_SHARED_CONTEXT;
+    info->p_info->caps = FI_RMA | FI_WRITE | FI_READ | /*SEND ONLY */
+        FI_ATOMICS; /* request atomics capability */
+    info->p_info->caps |= FI_REMOTE_WRITE | FI_REMOTE_READ;
+    info->p_info->tx_attr->op_flags = FI_DELIVERY_COMPLETE | FI_INJECT_COMPLETE;
+    info->p_info->mode = 0;
+    info->p_info->tx_attr->mode = 0;
+    info->p_info->rx_attr->mode = 0;
+
+    ctx->id = id;
+    shmem_internal_atomic_write(&ctx->pending_put_counter, 0);
+    shmem_internal_atomic_write(&ctx->pending_get_counter, 0);
+
+    ret = fi_cntr_open(shmem_transport_ofi_domainfd, &cntr_attr,
+                       &ctx->put_cntrfd, NULL);
+    if (ret!=0) {
+        RAISE_ERROR_MSG("context cntr_open failed (%s)\n", fi_strerror(errno));
+        return ret;
+    }
+
+    ret = fi_cntr_open(shmem_transport_ofi_domainfd, &cntr_attr,
+                       &ctx->get_cntrfd, NULL);
+    if (ret!=0) {
+        RAISE_ERROR_MSG("context cntr_open failed (%s)\n", fi_strerror(errno));
+        return ret;
+    }
+
+    ret = allocate_endpoints(ctx, info);
```

Comment:
I would pull the cntr_ep creation out of ```allocate_endpoints``` and do it here.  Then only use ```allocate_endpoints``` during initialization.